### PR TITLE
Initialize reference proxues with references to elements instead of l…

### PR DESCRIPTION
**Pull request by JohanMabille**
…ayout and index

### CODE REVIEW

File: include/sparrow/array_data.hpp [hunk 1]
```diff
@@ -412,7 +417,9 @@ namespace sparrow
     template <class U>
     void reference_proxy<L>::update_value(U&& u)
     {
-        p_layout->update(m_index, std::forward<U>(u));
+        m_bit_ref = true;
```
Comment (Klaim) [inline include/sparrow/array_data.hpp:420]: If the value is null, then `m_bit_ref = false` isnt it?

File: include/sparrow/dynamic_bitset.hpp [hunk 2]
```diff
@@ -309,6 +314,19 @@ namespace sparrow
         return m_null_count;
     }
 
+    template <random_access_range B>
+    auto dynamic_bitset_base<B>::operator[](size_type pos) -> reference
+    {
+        return reference(*this, m_buffer.data()[block_index(pos)], bit_mask(pos));
```
Comment (Klaim) [inline include/sparrow/dynamic_bitset.hpp:320]: Please add assertions about m_buffer's size vs pos :)

File: test/test_array_data.cpp [hunk 3]
```diff
@@ -25,62 +25,60 @@ namespace sparrow
...
         using reference = reference_proxy<mock_layout>;
         using const_reference = const_reference_proxy<mock_layout>;
         using size_type = std::size_t;
-        using storage_type = std::vector<value_type>;
 
         mock_layout() = default;
         mock_layout(std::initializer_list<value_type> l)
-            : m_storage(l)
+            : m_bitmap(l.size())
+            , m_data(l.size())
```
Comment (Klaim) [inline test/test_array_data.cpp:40]: So here we are default-constructing, not sure it's the best way to do this?

File: test/test_array_data.cpp [hunk 4]
```diff
@@ -25,62 +25,60 @@ namespace sparrow
...
+                {
+                    ++value_iter;
+                }
+            }
         }
 
         reference operator[](size_type pos)
         {
-            return reference(*this, pos);
+            return reference(m_data[pos], m_bitmap[pos]);
```
Comment (Klaim) [inline test/test_array_data.cpp:60]: Please add assertions about m_buffer's size vs pos :)

File: test/test_array_data.cpp [hunk 5]
```diff
@@ -25,62 +25,60 @@ namespace sparrow
...
-        {
-            return m_storage;
-        }
-
-    private:
-
-        bool has_value(size_type index) const
-        {
-            return m_storage[index].has_value();
+            return const_reference(m_data[pos], m_bitmap[pos]);
```
Comment (Klaim) [inline test/test_array_data.cpp:65]: Please add assertions about m_buffer's size vs pos :)

File: test/test_array_data.cpp [hunk 6]
```diff
@@ -25,62 +25,60 @@ namespace sparrow
...
         }
 
-        inner_value_type& value(size_type index)
+        bool has_value(size_type pos) const
         {
-            return m_storage[index].value();
+            return m_bitmap[pos];
```
Comment (Klaim) [inline test/test_array_data.cpp:70]: Please add assertions about m_buffer's size vs pos :)

File: test/test_array_data.cpp [hunk 7]
```diff
@@ -25,62 +25,60 @@ namespace sparrow
...
         }
 
-        const inner_value_type& value(size_type index) const
+        inner_const_reference value(size_type pos) const
         {
-            return m_storage[index].value();
+            return m_data[pos];
```
Comment (Klaim) [inline test/test_array_data.cpp:75]: Please add assertions about m_buffer's size vs pos :)

File: include/sparrow/array_data.hpp (hunk 1 above)
Comment (JohanMabille) [reply include/sparrow/array_data.hpp:420]: This method updates the inner value, not the value of the array (for this one, there is `update`)

File: test/test_array_data.cpp (hunk 3 above)
Comment (JohanMabille) [reply test/test_array_data.cpp:40]: I cannot think of a better alternative, but I'm opened to any suggestion ;)

**Review by jjerphan — APPROVED** [review]

File: test/test_array_data.cpp (hunk 3 above)
Comment (Klaim) [reply test/test_array_data.cpp:40]: welp, it's tests so I guess it's fine

**Pull request merged**